src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,733 @@
+     *
+     * @param operator the address to slash
+     * @param operatorSetId the ID of the operatorSet the operator is being slashed on behalf of
+     * @param strategies the set of strategies to slash
+     * @param bipsToSlash the number of bips to slash, this will be proportional to the
+     * operator's slashable stake allocation for the operatorSet
+     */
+    function slashOperator(
+        address operator,
+        uint32 operatorSetId,
+        IStrategy[] calldata strategies,
+        uint16 bipsToSlash
+    ) external {
+        OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: operatorSetId});
+        bytes32 operatorSetKey = _encodeOperatorSet(operatorSet);
+        require(isOperatorSlashable(operator, operatorSet), InvalidOperator());
+
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            // 1. slash the actively allocated (not pending deallocation) magnitude
+            uint64 slashedMagnitude;
+            {
+                Snapshots.History storage magnitudeUpdates = _magnitudeUpdate[operator][strategies[i]][operatorSetKey];
+                uint64 currentMagnitude = uint64(magnitudeUpdates.upperLookupRecent(uint32(block.timestamp)));
+                // skip if no actively allocated magnitude to slash
+                if (currentMagnitude != 0) {
+                    /// TODO: add wrapping library for rounding up for slashing accounting
+                    slashedMagnitude = uint64(uint256(bipsToSlash) * uint256(currentMagnitude) / BIPS_FACTOR);
+
+                    // propagate slashing to current and future and allocations for the (operator, strategy, operatorSet)
+                    magnitudeUpdates.decrementAtAndFutureSnapshots({
+                        key: uint32(block.timestamp),
+                        decrementValue: slashedMagnitude
+                    });
+                }
+            }
+
+            // 2. slash all pending deallocations that. pending deallocations are deallocations that are not completable yet
+            {
+                uint256 queuedDeallocationIndicesLen =
+                    _queuedDeallocationIndices[operator][strategies[i]][operatorSetKey].length;
+                for (uint256 j = queuedDeallocationIndicesLen; j > 0; --j) {
+                    // index of pendingFreeMagnitude/deallocation to check for slashing
+                    uint256 index = _queuedDeallocationIndices[operator][strategies[i]][operatorSetKey][j - 1];
+                    PendingFreeMagnitude storage pendingFreeMagnitude =
+                        _pendingFreeMagnitude[operator][strategies[i]][index];
+
+                    // Reached deallocation that is completable and, therefore, not slashable,
+                    // therefore older deallocations will also be completable. Since this is ordered by completableTimestamps break loop now
+                    if (block.timestamp >= pendingFreeMagnitude.completableTimestamp) {
```

Comment:
noting the edge case of equality here. I think having this be non-strict (as it is now) makes sense, but it would be good to clearly specify what the intended behavior is + make sure we are consistent about it in other places where strict vs. non-strict equality matters for this timespan -- e.g. when completing withdrawals

Reply:
We doubled down on limiting to a single pending deallocation so loop checking for each deallocation has been scrapped.

This is replaced with checks in `_getPendingMagnitudeInfo()` where a deallocation is NOT completable(is slashable) IFF `block.timestamp < effectTimestamp`.
Therefore a deallocation IS completable (not slashable) means `block.timestamp >= effectTimestamp`.

Now below is the check for completable timestamps for withdrawals in the DelegationManager
```
function getCompletableTimestamp(
        uint32 startTimestamp
    ) public view returns (uint32 completableTimestamp) {
        if (startTimestamp < LEGACY_WITHDRAWAL_CHECK_VALUE) {
            // this is a legacy M2 withdrawal using blocknumbers.
            // It would take 370+ years for the blockNumber to reach the LEGACY_WITHDRAWAL_CHECK_VALUE, so this is a safe check.
            require(startTimestamp + LEGACY_MIN_WITHDRAWAL_DELAY_BLOCKS <= block.number, WithdrawalDelayNotElapsed());
            // sourcing the magnitudes from time=0, will always give us WAD, which doesn't factor in slashing
            completableTimestamp = 0;
        } else {
            // this is a post Slashing release withdrawal using timestamps
            require(startTimestamp + MIN_WITHDRAWAL_DELAY <= block.timestamp, WithdrawalDelayNotElapsed());
            // source magnitudes from the time of completability
            completableTimestamp = startTimestamp + MIN_WITHDRAWAL_DELAY;
        }
    }
```
Symmetrically, we are checking here `startTimestamp + MIN_WITHDRAWAL_DELAY <= block.timestamp` which is correct as long as the immutable `MIN_WITHDRAWAL_DELAY` is configured to be the same as the `DEALLOCATION_DELAY` in the AllocationManager